src/Cassandra/Connection.cs

```diff
@@ -453,113 +471,171 @@ private void ReadHandler(byte[] buffer, int bytesReceived)
+                    _receivingHeader = header;
+                    break;
                 }
-                offset += headerSize;
-                count -= headerSize;
+                stream = stream ?? Configuration.BufferPool.GetStream(typeof (Connection) + "/Read");
+                OperationState state;
                 if (header.Opcode != EventResponse.OpCode)
                 {
-                    //Its a response to a previous request
-                    state = _pendingOperations[header.StreamId];
+                    state = RemoveFromPending(header.StreamId);
                 }
                 else
                 {
                     //Its an event
                     state = new OperationState(EventHandler);
                 }
-                state.Header = header;
-                _receivingOperation = state;
+                stream.Write(buffer, offset, remainingBodyLength);
+                var callback = state.SetCompleted();
+                operationCallbacks.AddLast(CreateCallbackAction(header, callback));
+                offset += remainingBodyLength;
             }
-            var countAdded = state.AppendBody(buffer, offset, count);
-
-            if (!state.IsBodyComplete)
+            if (operationCallbacks.Count == 0)
             {
-                //Nothing finished
+                //Not enough data to read a frame
                 return false;
             }
-            _logger.Verbose("Read #{0} for Opcode {1} from host {2}", state.Header.StreamId, state.Header.Opcode, Address);
-            //Stop reference it as the current receiving operation
-            _receivingOperation = null;
-            if (state.Header.Opcode != EventResponse.OpCode)
-            {
-                RemoveFromPending(state.Header.StreamId);
-            }
-            try
-            {
-                var response = ReadParseResponse(state.Header, state.BodyStream);
-                state.InvokeCallback(null, response);
-            }
-            catch (Exception ex)
+            //Invoke all callbacks using the default TaskScheduler
+            Task.Factory.StartNew(() =>
             {
-                state.InvokeCallback(ex);
-            }
-
-            if (countAdded < count)
-            {
-                //There is more data, from the next frame
-                ReadParse(buffer, offset + countAdded, count - countAdded);
-            }
+                // ReSharper disable once PossibleNullReferenceException
+                stream.Position = 0;
+                foreach (var cb in operationCallbacks)
+                {
+                    cb(stream);
+                }
+                stream.Dispose();
+            }, CancellationToken.None, TaskCreationOptions.None, TaskScheduler.Default);
             return true;
-            //There isn't enough data to read the whole frame.
-            //It is already buffered, carry on.
         }
```

Comment:
Ok, I know it is somewhat subjective but I think this method suffers from excessive complexity and excessive delegate usage. I would try to split the code in small methods and I would replace the `operationCallbacks` list with a list of `FrameToParse` that could be processed inside the `Task` action.
```cs
private class FrameToParse
{
    public readonly FrameHeader Header;
    public readonly Action<Exception, Response> StateCallback;
    // 
}
// replaces CreateCallbackAction, invoked inside the Task action,
private void ProcessFrame(FrameToParse frame, MemoryStream stream)
{
    //...
}
```